### ISSUE TITLE
Race condition in constructor when extending an Entity

### ISSUE THREAD

**Issue by eddieajau** [mateodelnorte/sourced#22]
I'm running into a bit of a roadblock when extending my Entity. It's compounded by TypeScript, but not unique to it. The problem is when extending object, TypeScript requires that the `super()` call (equivalent to `Entity.apply(this, arguments);`) comes as the first line in the `constructor` function. It's not possible to move the invocation such as can be found in [the example](https://github.com/mateodelnorte/sourced/blob/master/examples/auth/user.js#L5).

My musing has been store the data for the entity on a `state` property, the TypeScript will generate this kind of code:

``` js
    function EmailTemplateEntity() {
        _super.apply(this, arguments);
        this.state = new EmailTemplate();
    }
```

The issue is that when integrated with `sourced-repo-mongo`, the `Entity` constructor function is doing a lot of work before it even gets to my extended constructor. Hence, my `this.state` is not correctly initialised before the event methods in the entity are called.

Are you open to a fix for this (I generally try to avoid doing too much in constructors for this very reason)? The fix would require removing the merging of snapshots and replaying from the `Entity` constructor and then changing `Repository.prototype._deserialize` to invoke `Entity.merge` and `Entity.replay` explicitly.

Thanks in advance.

**Comment by mateodelnorte** [comment]
At this point, the API does not support using ES6 classes - specifically because the ECMAScript working group, in my opinion, made the horrible decision that in ES6 classes the `this` variable does not exist until you have called `super()`. They did this to make classes work in a way that is consistent with other object oriented languages, but in doing so they sacrificed the ability for a JS object to choose _when_ a parent constructor is fired. 

The result, for `sourced`, is that you currently have to write your classes using prototypical class structure.

**Comment by mateodelnorte** [comment]
I'm wearing of removing `merge` an `replay` from the constructor, however let me look over the code and reach out to some companies using `sourced` to make sure that won't break their usage.  Feel free to opine further while I do so.

**Comment by eddieajau** [comment]
> The result, for sourced, is that you currently have to write your classes using prototypical class structure.

Ok. I remember we touched on this. I've worked around it for now but we are committed to ES6 and beyond at work. I guess the alternative is that we can fork your work, experiment with things for a few months and then we can see if we can converge at a later date.

**Comment by mateodelnorte** [comment]
As a curiosity, are you using any of the static methods along with your ES6 classes?
```
.digestMethod(type, fn)
.mergeProperty(type, name, fn)
```

**Comment by eddieajau** [comment]
No, because `digestMethod` doesn't support `enqueue`. Here's an example:

```typescript
	public create(props: EmailTemplate) {
		props = _.omit(<EmailTemplate>{
			id: props.id,
			subject: props.subject,
			html: props.html,
			createdAt: props.createdAt || new Date(),
			createdBy: props.createdBy
		}, _.isUndefined);
		this.enqueue('created');
		this.merge(props);
		this.digest('create', props);
	}
```

**Comment by eddieajau** [comment]
I also don't use it because the magic around `this` makes it really, really difficult to write unit tests Netflix's Falcor does the same thing and it annoys me no end :) Yes, I know JavaScript lets you do things like that, but I don't like it because it you have to write yourself notes everywhere to remind you about what `this` is.

**Comment by mateodelnorte** [comment]
digestMethod supports .enqueue(). it takes at least two properties, same as emit. The purpose of .enqueue is to say "emit this event immediately after commit."

**Comment by eddieajau** [comment]
Sorry, I must have been thinking of something else then, possibly the service bus - so many moving parts now :)

**Comment by mateodelnorte** [comment]
Also, I wouldn't use Entity.merge() the way you are. Notice that's going to debug log text saying it's merging a snapshot. 

If you want to programmatically merge a number of properties to `this`, consider using something like https://lodash.com/docs/4.15.0#merge `merge(this, props);`. If you don't want all of lodash, you can import `lodash.merge`.

**Comment by eddieajau** [comment]
> consider using something like https://lodash.com/docs/4.15.0#merge merge(this, props);

Ok. Though, the version of lodash is quite old. Should that be upgraded?

> If you don't want all of lodash, you can import lodash.merge.

That's kind of moot because `Entity` loads the full library anyway?

**Comment by eddieajau** [comment]
> Ok. Though, the version of lodash is quite old. Should that be upgraded?

Let me rephrase. In order to not have to have 2 copies of lodash in my project, is it worth considering to upgrade the one in sourced to the latest version?

**Comment by mateodelnorte** [comment]
You can use whatever version of lodash you want in your application. But yea, that should probably be updated. I'll take a look later, as lodash has had some breaking changes over time.

**Comment by eddieajau** [comment]
> I'll take a look later, as lodash has had some breaking changes over time.

Yeah, that's what I was afraid of.

**Comment by patrickleet** [comment]
#26 es6 classes probably fix this

**Comment by patrickleet** [comment]
@eddieajau @mateodelnorte  closing this due to inactivity and due to the fact that ES6 classes in sourced 2.0 may address this. Feel free to reopen if it does not.
